File: regresion_lineal.py
# linear_regression.py
# Basado en el código de descenso de gradiente proporcionado
import numpy as np
import time
import psutil
import os

# Datos de ejemplo
X = np.array([1, 2, 3, 4, 5], dtype=float)
y = np.array([2, 4, 6, 8, 10], dtype=float)

# Inicialización de parámetros
w = 0.0  # pendiente
b = 0.0  # intercepto

# Hiperparámetros
learning_rate = 0.01
epochs = 1000

m = len(X)  # número de ejemplos

# Medir memoria antes del entrenamiento
process = psutil.Process(os.getpid())
mem_before = process.memory_info().rss / 1024 / 1024  # MB

# Medir tiempo
start_time = time.time()

# Entrenamiento con descenso de gradiente
for epoch in range(epochs):
    # Predicción del modelo: y_pred = w * X + b
    y_pred = w * X + b

    # Cálculo del error
    error = y_pred - y

    # Cálculo de gradientes (derivadas parciales)
    dw = (2/m) * np.dot(error, X)      # sum(error * X) * 2/m
    db = (2/m) * np.sum(error)          # sum(error) * 2/m

    # Actualización de parámetros
    w -= learning_rate * dw
    b -= learning_rate * db

    # (Opcional) Mostrar el costo cada cierto número de épocas
    if (epoch + 1) % 200 == 0:
        mse = np.mean(error ** 2)
        print(f"Epoch {epoch+1}, MSE: {mse:.4f}, w: {w:.4f}, b: {b:.4f}")

# Medir tiempo y memoria después del entrenamiento
end_time = time.time()
mem_after = process.memory_info().rss / 1024 / 1024  # MB

training_time = end_time - start_time
memory_used = mem_after - mem_before

print("\nModelo entrenado:")
print(f"w = {w:.4f}, b = {b:.4f}")

# Probar el modelo
x_nuevo = 7
y_pred_nuevo = w * x_nuevo + b
print(f"\nPara x = {x_nuevo}, y_pred = {y_pred_nuevo:.4f}")

# Mostrar métricas de desempeño
print(f"\n--- MÉTRICAS DE DESEMPEÑO ---")
print(f"Tiempo de entrenamiento: {training_time:.6f} segundos")
print(f"Memoria usada: {memory_used:.4f} MB")


# =====================================================
# VERSIÓN CON BENCHMARK COMPLETO
# =====================================================

def train_linear_regression(X, y, learning_rate=0.01, epochs=1000, verbose=False):
    """
    Entrena un modelo de regresión lineal usando descenso de gradiente.
    Retorna w, b, tiempo de entrenamiento y memoria usada.
    """
    w = 0.0
    b = 0.0
    m = len(X)

    # Medir memoria antes
    process = psutil.Process(os.getpid())
    mem_before = process.memory_info().rss / 1024 / 1024

    # Medir tiempo
    start_time = time.time()

    for epoch in range(epochs):
        # Predicción
        y_pred = w * X + b

        # Error
        error = y_pred - y

        # Gradientes
        dw = (2/m) * np.dot(error, X)
        db = (2/m) * np.sum(error)

        # Actualización
        w -= learning_rate * dw
        b -= learning_rate * db

        if verbose and (epoch + 1) % 200 == 0:
            mse = np.mean(error ** 2)
            print(f"Epoch {epoch+1}, MSE: {mse:.4f}")

    # Métricas finales
    end_time = time.time()
    mem_after = process.memory_info().rss / 1024 / 1024

    training_time = end_time - start_time
    memory_used = mem_after - mem_before

    return w, b, training_time, memory_used


def run_benchmark(n_samples_list=[100, 1000, 10000], n_runs=5):
    """
    Ejecuta benchmarks con diferentes tamaños de datasets.
    """
    print("="*70)
    print("BENCHMARK DE REGRESIÓN LINEAL - PYTHON")
    print("="*70)

    for n_samples in n_samples_list:
        print(f"\n### Dataset: {n_samples} muestras ###")
        print("-"*50)

        times = []
        memories = []

        for run in range(n_runs):
            # Generar datos sintéticos: y = 2*x + 3 + ruido
            np.random.seed(42 + run)
            X = np.random.randn(n_samples)
            y = 2 * X + 3 + np.random.randn(n_samples) * 0.1

            # Entrenar modelo
            w, b, train_time, mem_used = train_linear_regression(
                X, y, 
                learning_rate=0.01, 
                epochs=1000, 
                verbose=False
            )

            times.append(train_time)
            memories.append(mem_used)

            print(f"Run {run+1}: Tiempo={train_time:.6f}s, Memoria={mem_used:.4f}MB, w={w:.4f}, b={b:.4f}")

        # Estadísticas
        avg_time = np.mean(times)
        std_time = np.std(times)
        avg_memory = np.mean(memories)
        std_memory = np.std(memories)

        print("-"*50)
        print(f"PROMEDIO: Tiempo={avg_time:.6f}s, Memoria={avg_memory:.4f}MB")
        print(f"DESVIACIÓN: Tiempo={std_time:.6f}s, Memoria={std_memory:.4f}MB")


if __name__ == "__main__":
    print("\n### EJEMPLO BÁSICO (del código original) ###")
    print("="*70)

    # Ejecutar el ejemplo básico del código original
    X_simple = np.array([1, 2, 3, 4, 5], dtype=float)
    y_simple = np.array([2, 4, 6, 8, 10], dtype=float)

    w, b, train_time, mem_used = train_linear_regression(
        X_simple, y_simple, 
        learning_rate=0.01, 
        epochs=1000, 
        verbose=True
    )

    print(f"\nModelo final: w={w:.4f}, b={b:.4f}")
    print(f"Tiempo: {train_time:.6f}s, Memoria: {mem_used:.4f}MB")

    # Predicción
    x_nuevo = 7
    y_pred = w * x_nuevo + b
    print(f"\nPredicción para x={x_nuevo}: y={y_pred:.4f}")

    # Ejecutar benchmarks completos
    print("\n\n### BENCHMARKS COMPLETOS ###")
    run_benchmark(n_samples_list=[100, 1000, 10000], n_runs=5)

File: regresion_lineal.rs
// src/main.rs
// Implementación equivalente al código Python proporcionado
use std::time::Instant;

struct LinearRegressionModel {
    w: f64,  // pendiente
    b: f64,  // intercepto
}

impl LinearRegressionModel {
    fn new() -> Self {
        LinearRegressionModel { w: 0.0, b: 0.0 }
    }

    fn train(&mut self, x: &[f64], y: &[f64], learning_rate: f64, epochs: usize, verbose: bool) -> f64 {
        let m = x.len() as f64;
        let start = Instant::now();

        for epoch in 0..epochs {
            // Predicción: y_pred = w * X + b
            let y_pred: Vec<f64> = x.iter()
                .map(|&xi| self.w * xi + self.b)
                .collect();

            // Error: error = y_pred - y
            let error: Vec<f64> = y_pred.iter()
                .zip(y.iter())
                .map(|(pred, actual)| pred - actual)
                .collect();

            // Cálculo de gradientes (derivadas parciales)
            // dw = (2/m) * sum(error * X)
            let dw: f64 = (2.0 / m) * x.iter()
                .zip(error.iter())
                .map(|(xi, ei)| xi * ei)
                .sum::<f64>();

            // db = (2/m) * sum(error)
            let db: f64 = (2.0 / m) * error.iter().sum::<f64>();

            // Actualización de parámetros
            self.w -= learning_rate * dw;
            self.b -= learning_rate * db;

            // (Opcional) Mostrar el costo cada cierto número de épocas
            if verbose && (epoch + 1) % 200 == 0 {
                let mse: f64 = error.iter()
                    .map(|e| e * e)
                    .sum::<f64>() / m;
                println!("Epoch {}, MSE: {:.4}, w: {:.4}, b: {:.4}", 
                    epoch + 1, mse, self.w, self.b);
            }
        }

        let duration = start.elapsed();
        duration.as_secs_f64()
    }

    fn predict(&self, x: f64) -> f64 {
        self.w * x + self.b
    }
}

fn generate_data(n_samples: usize, seed: u64) -> (Vec<f64>, Vec<f64>) {
    use rand::Rng;
    use rand::SeedableRng;
    use rand_chacha::ChaCha8Rng;

    let mut rng = ChaCha8Rng::seed_from_u64(seed);

    let x: Vec<f64> = (0..n_samples)
        .map(|_| rng.gen_range(-3.0..3.0))
        .collect();

    let y: Vec<f64> = x.iter()
        .map(|&xi| 2.0 * xi + 3.0 + rng.gen_range(-0.1..0.1))
        .collect();

    (x, y)
}

fn run_benchmark(n_samples: usize, n_runs: usize) {
    println!("\n### Dataset: {} muestras ###", n_samples);
    println!("{}", "-".repeat(50));

    let mut times = Vec::new();

    for run in 0..n_runs {
        let (x, y) = generate_data(n_samples, 42 + run as u64);

        let mut model = LinearRegressionModel::new();
        let train_time = model.train(&x, &y, 0.01, 1000, false);

        times.push(train_time);

        println!("Run {}: Tiempo={:.6}s, w={:.4}, b={:.4}", 
            run + 1, train_time, model.w, model.b);
    }

    // Estadísticas
    let avg_time: f64 = times.iter().sum::<f64>() / times.len() as f64;
    let variance: f64 = times.iter()
        .map(|t| (t - avg_time).powi(2))
        .sum::<f64>() / times.len() as f64;
    let std_time = variance.sqrt();

    println!("{}", "-".repeat(50));
    println!("PROMEDIO: Tiempo={:.6}s", avg_time);
    println!("DESVIACIÓN: Tiempo={:.6}s", std_time);
}

fn main() {
    println!("### EJEMPLO BÁSICO (del código original) ###");
    println!("{}", "=".repeat(70));

    // Datos de ejemplo originales
    let x = vec![1.0, 2.0, 3.0, 4.0, 5.0];
    let y = vec![2.0, 4.0, 6.0, 8.0, 10.0];

    let mut model = LinearRegressionModel::new();
    let train_time = model.train(&x, &y, 0.01, 1000, true);

    println!("\nModelo final: w={:.4}, b={:.4}", model.w, model.b);
    println!("Tiempo: {:.6}s", train_time);

    // Predicción
    let x_nuevo = 7.0;
    let y_pred = model.predict(x_nuevo);
    println!("\nPredicción para x={}: y={:.4}", x_nuevo, y_pred);

    // Benchmarks completos
    println!("\n\n### BENCHMARKS COMPLETOS ###");
    println!("{}", "=".repeat(70));

    run_benchmark(100, 5);
    run_benchmark(1000, 5);
    run_benchmark(10000, 5);
}
